src/tree/updater_gpu_hist.cu

```diff
@@ -540,13 +551,19 @@ struct GPUHistMakerDevice {
     this->interaction_constraints.Reset();
     std::fill(node_sum_gradients.begin(), node_sum_gradients.end(),
               GradientPair());
+
+    if (use_gradient_based_sampling) {
+      auto sample = sampler->Sample(dh_gpair->DeviceSpan(), dmat);
+      n_rows = sample.sample_rows;
+      page = sample.page;
+      gpair = sample.gpair;
+    } else {
+      gpair = dh_gpair->DeviceSpan();
+      SubsampleGradientPair(device_id, gpair, param.subsample);
```

Comment:
this *may* alter gradient pairs. hence, if the caller is expecting this to be unaltered, this may result in a behavioral change. the boost methods are assuming this to be unaltered - more so, when there is more than a single updater. i'm not sure if such a configuration is practical - but, it looks to be there nonetheless.

my guess is that this is perhaps the reason why it is being copied in this specific tree updater.

Reply:
Probably should get rid of this in favor of the new gradient-based sampling.